File: dotnet/VM/StandardLibrary/StandardStringLibrary.cs
using System.Linq;
using System.Collections.Generic;

namespace LysitheaVM
{
    public static class StandardStringLibrary
    {
        #region Fields
        public static readonly IReadOnlyScope Scope = CreateScope();

        #endregion

        #region Methods
        public static Scope CreateScope()
        {
            var result = new Scope();

            var stringFunctions = new Dictionary<string, IValue>
            {
                {"length", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    vm.PushStack(new NumberValue(top.Value.Length));
                }, "string.length")},

                {"set", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var value = args.GetIndex(2);
                    vm.PushStack(Set(top, index.IntValue, value.ToString()));
                }, "string.set")},

                {"get", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    vm.PushStack(Get(top, index.IntValue));
                }, "string.get")},

                {"insert", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var value = args.GetIndex(2);
                    vm.PushStack(Insert(top, index.IntValue, value.ToString()));
                }, "string.insert")},

                {"substring", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var length = args.GetIndex<NumberValue>(2);
                    vm.PushStack(SubString(top, index.IntValue, length.IntValue));
                }, "string.substring")},

                {"removeAt", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    vm.PushStack(RemoveAt(top, index.IntValue));
                }, "string.removeAt")},

                {"removeAll", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<StringValue>(0);
                    var values = args.GetIndex<StringValue>(1);
                    vm.PushStack(RemoveAll(top, values.Value));
                }, "string.removeAll")},

                {"join", new BuiltinFunctionValue((vm, args) =>
                {
                    var separator = args.Value.First().ToString();
                    var result = string.Join(separator, args.Value.Skip(1));
                    vm.PushStack(result);
                }, "string.join")}
            };

            result.TryConstant("string", new ObjectValue(stringFunctions));

            return result;
        }

        public static StringValue Set(StringValue self, int index, string value)
        {
            index = self.GetIndex(index);
            var left = self.Value.Substring(0, index);
            var right = self.Value.Substring(index + 1);
            return new StringValue($"{left}{value}{right}");
        }

        public static StringValue Get(StringValue self, int index)
        {
            return new StringValue(self.Value[self.GetIndex(index)].ToString());
        }

        public static StringValue Insert(StringValue self, int index, string value)
        {
            return new StringValue(self.Value.Insert(self.GetIndex(index), value));
        }

        public static StringValue SubString(StringValue self, int index, int length)
        {
            return new StringValue(self.Value.Substring(self.GetIndex(index), length));
        }

        public static StringValue RemoveAt(StringValue self, int index)
        {
            return new StringValue(self.Value.Remove(self.GetIndex(index), 1));
        }

        public static StringValue RemoveAll(StringValue self, string values)
        {
            return new StringValue(self.Value.Replace(values, ""));
        }
        #endregion
    }
}

File: ts/src/standardLibrary/standardStringLibrary.ts
import { Scope, IReadOnlyScope } from "../scope";
import { BuiltinFunctionValue } from "../values/builtinFunctionValue";
import { ObjectValue, ObjectValueMap } from "../values/objectValue";
import { StringValue, isStringValue } from "../values/stringValue";

export const stringScope: IReadOnlyScope = createStringScope();

export function createStringScope()
{
    const result = new Scope();

    const stringFunctions: ObjectValueMap =
    {
        join: new BuiltinFunctionValue((vm, args) =>
        {
            if (args.value.length < 2)
            {
                throw new Error('Not enough arguments for string join');
            }

            const separator = args.value[0];
            const result = args.value.slice(1).join(separator.toString());
            vm.pushStackString(result);
        }, 'string.join'),

        length: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getString(0);
            vm.pushStackNumber(top.length);
        }, 'string.length'),

        get: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const index = args.getNumber(1);
            vm.pushStackString(top.value[top.getIndex(index)]);
        }, 'string.get'),

        set: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const index = args.getNumber(1);
            const value = args.getIndex(2).toString();
            vm.pushStack(set(top, index, value));
        }, 'string.set'),

        insert: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const index = args.getNumber(1);
            const value = args.getIndex(2).toString();
            vm.pushStack(insert(top, index, value));
        }, 'string.insert'),

        substring: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const index = args.getNumber(1);
            const length = args.getNumber(2);
            vm.pushStack(substring(top, index, length));
        }, 'string.substring'),

        removeAt: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const index = args.getNumber(1);
            vm.pushStack(removeAt(top, index));
        }, 'string.removeAt'),

        removeAll: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isStringValue);
            const values = args.getIndex(1).toString();
            vm.pushStack(removeAll(top, values));
        }, 'string.removeAll')
    }

    result.tryDefine('string', new ObjectValue(stringFunctions));

    return result;
}

export function set(input: StringValue, index: number, value: string): StringValue
{
    index = input.getIndex(index);
    return new StringValue(`${input.value.substring(0, index)}${value}${input.value.substring(index + 1)}`);
}

export function insert(input: StringValue, index: number, value: string): StringValue
{
    index = input.getIndex(index);
    return new StringValue(`${input.value.substring(0, index)}${value}${input.value.substring(index)}`);
}

export function removeAt(input: StringValue, index: number): StringValue
{
    index = input.getIndex(index);
    return new StringValue(`${input.value.substring(0, index)}${input.value.substring(index + 1)}`);
}

export function removeAll(input: StringValue, values: string): StringValue
{
    return new StringValue(input.value.replace(values, ''));
}

export function substring(input: StringValue, index: number, length: number): StringValue
{
    index = input.getIndex(index);
    return new StringValue(input.value.substring(index, index + length));
}
